File: Remove_Nth_node_from_end_of_list.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (head == NULL || head ->next == NULL){
            return NULL;
        }
        ListNode *current = head;
        for (int i=0; i<n; i++){
            current = current ->next;
        }
        ListNode *prev = head;
        if (current == NULL){
            head = prev ->next;
            return head;
        }
        while (current ->next != NULL){
            current = current ->next;
            prev = prev ->next;
        }
        prev ->next = prev ->next ->next;
        return head;
    }
};

File: new_version/remove_nth_node_from_end_of_list.rb
# =====================
# jump n step from head and remember this node with tail
# jump head and tail together until tail reach nil(end of list)
# remove current head and return head of list will be okay
# =====================

# Definition for singly-linked list.
 class ListNode
	 attr_accessor :val, :next
	 def initialize(val)
		 @val = val
		 @next = nil
	 end
 end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
	tail = head
	prev = head
	while n > 0
		return nil if tail == nil
		tail = tail.next
		n -= 1
	end
	return head.next if tail == nil
	tail = tail.next
	while tail != nil
		tail = tail.next
		prev = prev.next
	end
	prev.next = prev.next.next
	head
end

def arr2list(arr)
	temp = nil
	arr.reverse.each do |e|
		current = ListNode.new e
		current.next = temp
		temp = current
	end
	temp
end

l = [2]
puts remove_nth_from_end(arr2list(l), 2).inspect
